src/test/java/com/powsybl/openloadflow/ac/AcLoadFlowTransformerControlTest.java

```diff
@@ -0,0 +1,215 @@
+                .setR(17.0)
+                .setX(10.0)
+                .setG(0.00573921028466483)
+                .setB(0.000573921028466483)
+                .setBus1("BUS_2")
+                .setBus2("BUS_3")
+                .add();
+
+        twt.newRatioTapChanger()
+                .beginStep()
+                .setRho(0.9)
+                .setR(0.1089)
+                .setX(0.01089)
+                .setG(0.8264462809917356)
+                .setB(0.08264462809917356)
+                .endStep()
+                .beginStep()
+                .setRho(1.0)
+                .setR(0.121)
+                .setX(0.0121)
+                .setG(0.8264462809917356)
+                .setB(0.08264462809917356)
+                .endStep()
+                .beginStep()
+                .setRho(1.1)
+                .setR(0.1331)
+                .setX(0.01331)
+                .setG(0.9090909090909092)
+                .setB(0.09090909090909092)
+                .endStep()
+                .setTapPosition(0)
+                .setLoadTapChangingCapabilities(true)
+                .setRegulating(false)
+                .setTargetV(33.0)
+                .setRegulationTerminal(load3.getTerminal())
+                .add();
+
+        return network;
+    }
+
+    @BeforeEach
+    void setUp() {
+
+        network = createNetwork();
+        loadFlowRunner = new LoadFlow.Runner(new OpenLoadFlowProvider(new DenseMatrixFactory()));
+        parameters = new LoadFlowParameters();
+        parameters.setTransformerVoltageControlOn(false);
+        OpenLoadFlowParameters parametersExt = new OpenLoadFlowParameters()
+                .setSlackBusSelector(new FirstSlackBusSelector())
+                .setDistributedSlack(false);
+        this.parameters.addExtension(OpenLoadFlowParameters.class, parametersExt);
+    }
+
+    @Test
+    void baseCaseTest() {
+        LoadFlowResult result = loadFlowRunner.run(network, parameters);
+        assertTrue(result.isOk());
+
+        assertVoltageEquals(135.0, bus1);
+        assertVoltageEquals(134.273, bus2);
+        assertVoltageEquals(27.0038, bus3);
+    }
+
+    @Test
+    void tapPlusTwoTest() {
+        twt.getRatioTapChanger().setTapPosition(2);
+        LoadFlowResult result = loadFlowRunner.run(network, parameters);
+        assertTrue(result.isOk());
+
+        assertVoltageEquals(135.0, bus1);
+        assertVoltageEquals(134.281, bus2);
+        assertVoltageEquals(34.427, bus3);
+    }
+
+    @Test
+    void voltageControlTest() {
+        parameters.setTransformerVoltageControlOn(true);
+        twt.getRatioTapChanger()
+                .setTargetDeadband(0)
+                .setRegulating(true)
+                .setTapPosition(0)
+                .setRegulationTerminal(twt.getTerminal2())
+                .setTargetV(34.0);
+
+        LoadFlowResult result = loadFlowRunner.run(network, parameters);
+        assertTrue(result.isOk());
+        assertVoltageEquals(34.433, bus3);
```

Comment:
@geofjamg I have a small difference but I am not sure to understand why...

Reply:
Because we round the r1?

Reply:
There is something missing in the algo, maybe after rounding and fixing ratio it is to generation behind transformers du help reaching the target voltage?